File: shared/src/main/java/ru/privatenull/pnauth/policy/AccessSettings.java
package ru.privatenull.pnauth.policy;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public record AccessSettings(
        boolean blockChat,
        Set<String> unauthenticatedCommands
) {
    public AccessSettings {
        unauthenticatedCommands = unauthenticatedCommands == null
                ? Set.of()
                : Set.copyOf(normalize(unauthenticatedCommands));
    }

    public static AccessSettings defaults() {
        return new AccessSettings(
                true,
                Set.of("auth", "pnauth", "register", "reg", "login", "l", "logout",
                        "changepassword", "changepass", "totp", "2fa", "status")
        );
    }

    private static Set<String> normalize(Set<String> commands) {
        Set<String> normalized = new LinkedHashSet<>();
        for (String command : commands) {
            if (command != null && !command.isBlank()) {
                normalized.add(command.trim().toLowerCase(Locale.ROOT).replaceFirst("^/", ""));
            }
        }
        return normalized;
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/policy/AccessSettings.kt
package ru.privatenull.pnauth.policy

import java.util.Locale

class AccessSettings @JvmOverloads constructor(
    val blockChat: Boolean,
    unauthenticatedCommands: Set<String>? = emptySet()
) {
    val unauthenticatedCommands: Set<String> = normalize(unauthenticatedCommands)

    fun blockChat(): Boolean = blockChat
    fun unauthenticatedCommands(): Set<String> = unauthenticatedCommands

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AccessSettings) return false
        return blockChat == other.blockChat && unauthenticatedCommands == other.unauthenticatedCommands
    }

    override fun hashCode(): Int {
        return 31 * blockChat.hashCode() + unauthenticatedCommands.hashCode()
    }

    override fun toString(): String {
        return "AccessSettings(blockChat=$blockChat, unauthenticatedCommands=$unauthenticatedCommands)"
    }

    companion object {
        @JvmStatic
        fun defaults(): AccessSettings = AccessSettings(
            true,
            setOf(
                "auth", "pnauth", "register", "reg", "login", "l", "logout",
                "changepassword", "changepass", "totp", "2fa", "status"
            )
        )

        @JvmStatic
        fun normalize(commands: Set<String>?): Set<String> {
            if (commands == null) return emptySet()
            val normalized = LinkedHashSet<String>()
            for (command in commands) {
                if (command.isNotBlank()) {
                    normalized.add(command.trim().lowercase(Locale.ROOT).replaceFirst("^/".toRegex(), ""))
                }
            }
            return java.util.Set.copyOf(normalized)
        }
    }
}
